### PR TITLE
MM-57839: Rewrite static assets only if needed

### LINKED ISSUE #26769: root.html updated even when not needed
#### Summary
root.html updated even when not needed every time on server start

```
{"timestamp":"2024-04-12 01:05:57.990 -04:00","level":"debug","msg":"Rewriting static assets","caller":"utils/subpath.go:83","from_subpath":"/","to_subpath":"/"}
```
Note, both `from_subpath` and `to_subpath` are same `/`. Yet it tries to write the file

#### Steps to reproduce
Version: 9.5.3
Enable debug logging and restart server.

#### Expected behavior
No write to client directory. So should be possible to keep it readonly.

#### Possible fixes
Check if old and new values are different and only then attempt to write. The comment here is wrong.
https://github.com/mattermost/mattermost/blob/v9.5.3/server/channels/utils/subpath.go#L48

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary
Refactor `UpdateAssetsSubpathInDir` so that the actual rewriting of files happen in two different functions: one for `root.html`, another for `manifest.json` and `*.css` files.

I would have wanted to simply do an `if pathToReplace == newPath { return nil }` when those two are defined, but the logic for `root.html` is not that simple*, and that may miss some edge cases, so I opted out for simply moving the actual rewriting to individual functions, which check for their corresponding conditions to perform the update:
1. for `root.html`, check that the edited file is indeed different than the original one
2. for `manifest.json` and `*.css` files, rewrite them only if `pathToReplace != newPath`, which in this case is clear that that's the only modification we do.

The logic is exactly the same as before except for skipping the rewrite when those conditions are not met.

*I don't really understand this block:
```go
// Determine if a previous subpath had already been rewritten into the assets.
reWebpackPublicPathScript := regexp.MustCompile("window.publicPath='([^']+/)static/'")
alreadyRewritten := false
if matches := reWebpackPublicPathScript.FindStringSubmatch(string(oldRootHTML)); matches != nil {
	oldSubpath = matches[1]
	alreadyRewritten = true
}
```
When can that be rewritten if not here? 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-57839

#### Screenshots
--

#### Release Note
```release-note
NONE
```

Fix #26769

### CODE REVIEW

**Comment by agarciamontoro** [comment]
@hmhealey, adding you here in case you're familiar with this client static assets rewriting.

**Comment by agnivade** [comment]
Swapping me for Jesse, who should have more context on this area.

File: server/channels/utils/subpath.go [hunk 1]
```diff
@@ -110,13 +124,29 @@ func UpdateAssetsSubpathInDir(subpath, directory string) error {
 		newRootHTML = strings.Replace(newRootHTML, "</style>", fmt.Sprintf("</style><script>%s</script>", script), 1)
 	}
 
+	if newRootHTML == oldRootHTML {
+		mlog.Debug("No need to rewrite unmodified root.html", mlog.String("from_subpath", pathToReplace), mlog.String("to_subpath", newPath))
```
Comment (lieut-data) [inline server/channels/utils/subpath.go:128]: Would it make sense to just log nothing in this case? (Unix philosophy and all?)

**Review by lieut-data — APPROVED** [review]
Nice!

File: server/channels/utils/subpath.go (hunk 1 above)
Comment (agarciamontoro) [reply server/channels/utils/subpath.go:128]: Oh, I didn't know that was part of the Unix philosophy :eyes: But given this is a debug-level log, I'm in favour of keeping it. I don't think there's such thing as a lot of debug logging! If you have a stronger opinion, I can change it.

Comment (lieut-data) [reply server/channels/utils/subpath.go:128]: I'm only 0/5, but if we DEBUG logged everything that /didn't/ happen, I would wager that would be a /lot/ of debug logging ;P. My hope whenever scrubbing logs is to look for positive evidence of something, and in that context, fewer logs that speak to something that actually happened are strictly more helpful. But this is a one-off, so I'm really only -1/5 :D

**Comment by agarciamontoro** [comment]
@DHaussermann, this PR modifies the code that updates the static assets served by MM when changing the subpath where the application is served. Could you run a smoke test to ensure that everything is working as expected? If something is broken, it should be quite obvious, as in nothing works. We should ideally test that changing the subpath also works, but I don't know how easy that is (if it's even possible) in a cloud test server.

**Review by DHaussermann — APPROVED** [review]
Tested and Passed
- Populated browser's application data with assets on the subpath server
- Had the server restarted where proxy was now serving me the app on a new subpath and site URL had been adjusted accordingly
- Explored the app - no functional issues found
- No relevant issues found in JS console or Network tab
- Logs showed `"msg":"Rewriting root.html"` and `Rewriting manifest.json and *.css files` as expected
- As a precaution explored the app again with fresh cache. 
No issues found.

LGTM!

Huge thanks to @toninis for the infrastructure help on this one 🎉

**Comment by agarciamontoro** [comment]
Thank you, @DHaussermann, and thank you, @toninis <3

**Pull request merged**